File: app/src/main/java/com/krystianwsul/checkme/persistencemodel/TaskRecord.java
package com.krystianwsul.checkme.persistencemodel;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import junit.framework.Assert;

import java.util.ArrayList;
import java.util.List;

public class TaskRecord extends Record {
    static final String TABLE_TASKS = "tasks";

    static final String COLUMN_ID = "_id";
    private static final String COLUMN_NAME = "name";
    private static final String COLUMN_START_TIME = "startTime";
    private static final String COLUMN_END_TIME = "endTime";
    private static final String COLUMN_OLDEST_VISIBLE_YEAR = "oldestVisibleYear";
    private static final String COLUMN_OLDEST_VISIBLE_MONTH = "oldestVisibleMonth";
    private static final String COLUMN_OLDEST_VISIBLE_DAY = "oldestVisibleDay";
    private static final String COLUMN_NOTE = "note";

    private final int id;

    @NonNull
    private String name;

    private final long startTime;

    @Nullable
    private Long endTime;

    @Nullable
    private Integer oldestVisibleYear;

    @Nullable
    private Integer oldestVisibleMonth;

    @Nullable
    private Integer oldestVisibleDay;

    @Nullable
    private String note;

    public static void onCreate(@NonNull SQLiteDatabase sqLiteDatabase) {
        sqLiteDatabase.execSQL("CREATE TABLE " + TABLE_TASKS
                + " (" + COLUMN_ID + " INTEGER PRIMARY KEY AUTOINCREMENT, "
                + COLUMN_NAME + " TEXT NOT NULL, "
                + COLUMN_START_TIME + " TEXT NOT NULL, "
                + COLUMN_END_TIME + " TEXT, "
                + COLUMN_OLDEST_VISIBLE_YEAR + " INTEGER, "
                + COLUMN_OLDEST_VISIBLE_MONTH + " INTEGER, "
                + COLUMN_OLDEST_VISIBLE_DAY + " INTEGER, "
                + COLUMN_NOTE + " TEXT);");
    }

    @NonNull
    static List<TaskRecord> getTaskRecords(@NonNull SQLiteDatabase sqLiteDatabase) {
        ArrayList<TaskRecord> taskRecords = new ArrayList<>();

        Cursor cursor = sqLiteDatabase.query(TABLE_TASKS, null, null, null, null, null, null);
        cursor.moveToFirst();
        while (!cursor.isAfterLast()) {
            taskRecords.add(cursorToTaskRecord(cursor));
            cursor.moveToNext();
        }
        cursor.close();

        return taskRecords;
    }

    @NonNull
    private static TaskRecord cursorToTaskRecord(@NonNull Cursor cursor) {
        int id = cursor.getInt(0);
        String name = cursor.getString(1);
        long startTime = cursor.getLong(2);
        Long endTime = (cursor.isNull(3) ? null : cursor.getLong(3));
        Integer oldestVisibleYear = (cursor.isNull(4) ? null : cursor.getInt(4));
        Integer oldestVisibleMonth = (cursor.isNull(5) ? null : cursor.getInt(5));
        Integer oldestVisibleDay = (cursor.isNull(6) ? null : cursor.getInt(6));
        String note = cursor.getString(7);

        Assert.assertTrue(!TextUtils.isEmpty(name));
        Assert.assertTrue(endTime == null || startTime <= endTime);
        Assert.assertTrue((oldestVisibleYear == null) == (oldestVisibleMonth == null));
        Assert.assertTrue((oldestVisibleYear == null) == (oldestVisibleDay == null));

        return new TaskRecord(true, id, name, startTime, endTime, oldestVisibleYear, oldestVisibleMonth, oldestVisibleDay, note);
    }

    static int getMaxId(@NonNull SQLiteDatabase sqLiteDatabase) {
        return getMaxId(sqLiteDatabase, TABLE_TASKS, COLUMN_ID);
    }

    TaskRecord(boolean created, int id, @NonNull String name, long startTime, @Nullable Long endTime, @Nullable Integer oldestVisibleYear, @Nullable Integer oldestVisibleMonth, @Nullable Integer oldestVisibleDay, @Nullable String note) {
        super(created);

        Assert.assertTrue(!TextUtils.isEmpty(name));
        Assert.assertTrue(endTime == null || startTime <= endTime);
        Assert.assertTrue((oldestVisibleYear == null) == (oldestVisibleMonth == null));
        Assert.assertTrue((oldestVisibleYear == null) == (oldestVisibleDay == null));

        this.id = id;
        this.name = name;
        this.startTime = startTime;
        this.endTime = endTime;

        this.oldestVisibleDay = oldestVisibleDay;
        this.oldestVisibleMonth = oldestVisibleMonth;
        this.oldestVisibleYear = oldestVisibleYear;

        this.note = note;
    }

    public int getId() {
        return id;
    }

    @NonNull
    public String getName() {
        return name;
    }

    public long getStartTime() {
        return startTime;
    }

    @Nullable
    public Long getEndTime() {
        return endTime;
    }

    @Nullable
    public Integer getOldestVisibleYear() {
        return oldestVisibleYear;
    }

    @Nullable
    public Integer getOldestVisibleMonth() {
        return oldestVisibleMonth;
    }

    @Nullable
    public Integer getOldestVisibleDay() {
        return oldestVisibleDay;
    }

    @Nullable
    public String getNote() {
        return note;
    }

    public void setName(@Nullable String name) {
        Assert.assertTrue(!TextUtils.isEmpty(name));

        this.name = name;
        changed = true;
    }

    public void setEndTime(long endTime) {
        Assert.assertTrue(this.endTime == null);
        Assert.assertTrue(startTime <= endTime);

        this.endTime = endTime;
        changed = true;
    }

    public void setOldestVisibleYear(int oldestVisibleYear) {
        this.oldestVisibleYear = oldestVisibleYear;
        changed = true;
    }

    public void setOldestVisibleMonth(int oldestVisibleYearMonth) {
        oldestVisibleMonth = oldestVisibleYearMonth;
        changed = true;
    }

    public void setOldestVisibleDay(int oldestVisibleDay) {
        this.oldestVisibleDay = oldestVisibleDay;
        changed = true;
    }

    public void setNote(@Nullable String note) {
        this.note = note;
        changed = true;
    }

    @NonNull
    @Override
    ContentValues getContentValues() {
        ContentValues contentValues = new ContentValues();
        contentValues.put(COLUMN_NAME, name);
        contentValues.put(COLUMN_START_TIME, startTime);
        contentValues.put(COLUMN_END_TIME, endTime);
        contentValues.put(COLUMN_OLDEST_VISIBLE_YEAR, oldestVisibleYear);
        contentValues.put(COLUMN_OLDEST_VISIBLE_MONTH, oldestVisibleMonth);
        contentValues.put(COLUMN_OLDEST_VISIBLE_DAY, oldestVisibleDay);
        contentValues.put(COLUMN_NOTE, note);
        return contentValues;
    }

    @NonNull
    @Override
    UpdateCommand getUpdateCommand() {
        return getUpdateCommand(TABLE_TASKS, COLUMN_ID, id);
    }

    @NonNull
    @Override
    InsertCommand getInsertCommand() {
        return getInsertCommand(TABLE_TASKS);
    }

    @NonNull
    @Override
    DeleteCommand getDeleteCommand() {
        return getDeleteCommand(TABLE_TASKS, COLUMN_ID, id);
    }
}

File: app/src/main/java/com/krystianwsul/checkme/persistencemodel/TaskRecord.kt
package com.krystianwsul.checkme.persistencemodel

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase

import kotlin.properties.Delegates.observable

class TaskRecord(created: Boolean, val id: Int, _name: String, val startTime: Long, _endTime: Long?, _oldestVisibleYear: Int?, _oldestVisibleMonth: Int?, _oldestVisibleDay: Int?, _note: String?) : Record(created) {

    companion object {

        val TABLE_TASKS = "tasks"

        val COLUMN_ID = "_id"
        private const val COLUMN_NAME = "name"
        private const val COLUMN_START_TIME = "startTime"
        private const val COLUMN_END_TIME = "endTime"
        private const val COLUMN_OLDEST_VISIBLE_YEAR = "oldestVisibleYear"
        private const val COLUMN_OLDEST_VISIBLE_MONTH = "oldestVisibleMonth"
        private const val COLUMN_OLDEST_VISIBLE_DAY = "oldestVisibleDay"
        private const val COLUMN_NOTE = "note"

        fun onCreate(sqLiteDatabase: SQLiteDatabase) {
            sqLiteDatabase.execSQL("CREATE TABLE $TABLE_TASKS " +
                    "($COLUMN_ID INTEGER PRIMARY KEY AUTOINCREMENT, " +
                    "$COLUMN_NAME TEXT NOT NULL, " +
                    "$COLUMN_START_TIME TEXT NOT NULL, " +
                    "$COLUMN_END_TIME TEXT, " +
                    "$COLUMN_OLDEST_VISIBLE_YEAR INTEGER, " +
                    "$COLUMN_OLDEST_VISIBLE_MONTH INTEGER, " +
                    "$COLUMN_OLDEST_VISIBLE_DAY INTEGER, " +
                    "$COLUMN_NOTE TEXT);")
        }

        fun getTaskRecords(sqLiteDatabase: SQLiteDatabase) = getRecords(sqLiteDatabase, TABLE_TASKS, this::cursorToTaskRecord)

        private fun cursorToTaskRecord(cursor: Cursor) = cursor.run {
            val id = getInt(0)
            val name = getString(1)
            val startTime = getLong(2)
            val endTime = if (isNull(3)) null else getLong(3)
            val oldestVisibleYear = if (isNull(4)) null else getInt(4)
            val oldestVisibleMonth = if (isNull(5)) null else getInt(5)
            val oldestVisibleDay = if (isNull(6)) null else getInt(6)
            val note = getString(7)

            check(name.isNotEmpty())
            check(endTime == null || startTime <= endTime)
            check(oldestVisibleYear == null == (oldestVisibleMonth == null))
            check(oldestVisibleYear == null == (oldestVisibleDay == null))

            TaskRecord(true, id, name, startTime, endTime, oldestVisibleYear, oldestVisibleMonth, oldestVisibleDay, note)
        }

        fun getMaxId(sqLiteDatabase: SQLiteDatabase) = Record.getMaxId(sqLiteDatabase, TABLE_TASKS, COLUMN_ID)
    }

    var name by observable(_name) { _, _, newValue ->
        check(newValue.isNotEmpty())

        changed = true
    }

    var endTime by observable(_endTime) { _, _, newValue ->
        newValue?.let { check(startTime <= it) }

        changed = true
    }

    var oldestVisibleYear by observable(_oldestVisibleYear) { _, _, _ -> changed = true }

    var oldestVisibleMonth by observable(_oldestVisibleMonth) { _, _, _ -> changed = true }

    var oldestVisibleDay by observable(_oldestVisibleDay) { _, _, _ -> changed = true }

    var note by observable(_note) { _, _, _ -> changed = true }

    init {
        check(_name.isNotEmpty())
        check(_endTime == null || startTime <= _endTime)
        check(_oldestVisibleYear == null == (_oldestVisibleMonth == null))
        check(_oldestVisibleYear == null == (_oldestVisibleDay == null))
    }

    override val contentValues
        get() = ContentValues().apply {
        put(COLUMN_NAME, name)
        put(COLUMN_START_TIME, startTime)
        put(COLUMN_END_TIME, endTime)
        put(COLUMN_OLDEST_VISIBLE_YEAR, oldestVisibleYear)
        put(COLUMN_OLDEST_VISIBLE_MONTH, oldestVisibleMonth)
        put(COLUMN_OLDEST_VISIBLE_DAY, oldestVisibleDay)
        put(COLUMN_NOTE, note)
    }

    override val commandTable = TABLE_TASKS
    override val commandIdColumn = COLUMN_ID
    override val commandId = id
}
